server/blob/blob-deduplicating/src/main/scala/GC.scala

```diff
@@ -0,0 +1,192 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.server.blob.deduplication
+
+import org.apache.james.blob.api.BlobId
+
+sealed abstract class Generation extends Comparable[Generation] {
+  def previous: Generation
+  def previous(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.previous)
+
+  def next: Generation
+  def next(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.next)
+
+  def collectibles(targetedGeneration: Generation): Set[Generation] =
+    Generation.range(next, targetedGeneration.previous(GC.temporization)).toSet
+
+  def <(that : Generation): Boolean = compareTo(that) < 0
+  def <=(that : Generation): Boolean = compareTo(that) <= 0
+  def >(that : Generation): Boolean = compareTo(that) > 0
+  def >=(that : Generation): Boolean = compareTo(that) >= 0
+}
+
+case class ValidGeneration(id: Long) extends Generation {
+  override def previous: Generation = Generation(id - 1)
+
+  override def next: Generation = copy(id + 1)
+
+  override def compareTo(t: Generation): Int = t match {
+    case InvalidGeneration => 1
+    case that : ValidGeneration => id.compareTo(that.id)
+  }
+
+}
+
+case object InvalidGeneration extends Generation {
+  override def previous: Generation = InvalidGeneration
+
+  override def next: Generation = Generation.first
+
+  override def compareTo(t: Generation): Int = t match {
+    case InvalidGeneration => 0
+    case _ : ValidGeneration => -1
+  }
+}
+
+object Generation {
+  val first: Generation = apply(0)
+
+  def apply(id: Long): Generation = {
+    if (id < 0) {
+      InvalidGeneration
+    } else {
+      ValidGeneration(id)
+    }
+  }
+
+  def range(start : Generation, end : Generation) : Seq[Generation] = (start, end) match {
+    case (InvalidGeneration, InvalidGeneration) => Seq(InvalidGeneration)
+    case (ValidGeneration(_), InvalidGeneration) => Nil
+    case (InvalidGeneration, ValidGeneration(id)) =>  InvalidGeneration +: (0L to id).map(Generation.apply)
+    case (ValidGeneration(id1), ValidGeneration(id2)) => (id1 to id2).map(Generation.apply)
+  }
+}
+
+case class Iteration(id: Long, processedGenerations: Set[Generation], lastGeneration: Generation) {
+  def next(generations: Set[Generation], lastGeneration: Generation): Iteration = Iteration(id + 1, generations, lastGeneration)
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(InvalidGeneration), InvalidGeneration)
+}
+
+case class ExternalID(id: String) // TODO
+
+sealed trait Event {
+  def blob: BlobId
+  def externalId: ExternalID
+  def generation: Generation
```

Comment:
So the generation is external to the blobId right?

The BlobId  structure is unchanged, and the references to that blob id are scoped by generation, right?

If so I fear a given blobId could be referenced by several generations (even the newest) and thus we have no concurrency safety.

For us to be concurrency safe we need to have the generation being a subpart of the blob Id.

```
The de-duplicating id of the blobs which before where constructed
using only the hash of their content,  will now include this `reference generation` too.`
```

(unmerged ADR btw you are asking for trouble) https://github.com/linagora/james-project/pull/3122/files

What I understand is that we need a `DeduplicationBlobId` that composes a `HashBlobId` and a `Generation` but we cannot have two distinct fields.

Reply:
We currently rely on an specific test `BlobId`, which will be what we will do within the DumbBlobStore work.